internal/ls/folding.go

```diff
@@ -0,0 +1,541 @@
+	p := &printer.EmitContext{}
+	comments := scanner.GetLeadingCommentRanges(&printer.NewNodeFactory(p).NodeFactory, sourceFile.Text(), pos)
+
+	var foldingRange []*lsproto.FoldingRange
+	firstSingleLineCommentStart := -1
+	lastSingleLineCommentEnd := -1
+	singleLineCommentCount := 0
+	foldingRangeKindComment := lsproto.FoldingRangeKindComment
+
+	combineAndAddMultipleSingleLineComments := func() *lsproto.FoldingRange {
+		// Only outline spans of two or more consecutive single line comments
+		if singleLineCommentCount > 1 {
+			return createFoldingRangeFromBounds(firstSingleLineCommentStart, lastSingleLineCommentEnd, foldingRangeKindComment, sourceFile, l)
+		}
+		return nil
+	}
+
+	sourceText := sourceFile.Text()
+	comments(func(comment ast.CommentRange) bool {
+		commentPos := comment.Pos()
+		commentEnd := comment.End()
+		// cancellationToken.throwIfCancellationRequested();
+		switch comment.Kind {
+		case ast.KindSingleLineCommentTrivia:
+			// never fold region delimiters into single-line comment regions
+			commentText := sourceText[commentPos:commentEnd]
+			if parseRegionDelimiter(commentText) != nil {
+				comments := combineAndAddMultipleSingleLineComments()
+				if comments != nil {
+					foldingRange = append(foldingRange, comments)
+				}
+				singleLineCommentCount = 0
+				break
+			}
+
+			// For single line comments, combine consecutive ones (2 or more) into
+			// a single span from the start of the first till the end of the last
+			if singleLineCommentCount == 0 {
+				firstSingleLineCommentStart = commentPos
+			}
+			lastSingleLineCommentEnd = commentEnd
+			singleLineCommentCount++
+			break
+		case ast.KindMultiLineCommentTrivia:
+			comments := combineAndAddMultipleSingleLineComments()
+			if comments != nil {
+				foldingRange = append(foldingRange, comments)
+			}
+			foldingRange = append(foldingRange, createFoldingRangeFromBounds(commentPos, commentEnd, foldingRangeKindComment, sourceFile, l))
+			singleLineCommentCount = 0
+			break
+		default:
+			// Debug.assertNever(kind);
+		}
+		return true
+	})
+	addedComments := combineAndAddMultipleSingleLineComments()
+	if addedComments != nil {
+		foldingRange = append(foldingRange, addedComments)
+	}
+	return foldingRange
+}
+
+var regionDelimiterRegExp = regexp.MustCompile(`^#(end)?region(.*)\r?$`)
+
+type regionDelimiterResult struct {
+	isStart bool
+	name    string
+}
+
+func parseRegionDelimiter(lineText string) *regionDelimiterResult {
+	// We trim the leading whitespace and // without the regex since the
+	// multiple potential whitespace matches can make for some gnarly backtracking behavior
+	lineText = strings.TrimLeftFunc(lineText, unicode.IsSpace)
+	if !strings.HasPrefix(lineText, "//") {
+		return nil
+	}
+	lineText = strings.TrimSpace(lineText[2:])
+	result := regionDelimiterRegExp.FindStringSubmatch(lineText)
+	if result != nil {
+		return &regionDelimiterResult{
```

Comment:
@jakebailey it's minor, but any idea if the allocation can be eliminated here? I might just make `parseRegionDelimiter` return `result, ok`, but I get wanting to preserve the original code.